chain/client/src/chunk_inclusion_tracker.rs

```diff
@@ -1,37 +1,60 @@
+use chrono::{DateTime, Utc};
 use itertools::Itertools;
 use lru::LruCache;
+use near_primitives::checked_feature;
+use near_primitives::version::ProtocolVersion;
 use std::collections::HashMap;
+use std::sync::Arc;
 
+use near_chain_primitives::Error;
+use near_epoch_manager::EpochManagerAdapter;
+use near_primitives::block_body::ChunkEndorsementSignatures;
 use near_primitives::hash::CryptoHash;
-use near_primitives::sharding::ShardChunkHeader;
+use near_primitives::sharding::{ChunkHash, ShardChunkHeader};
 use near_primitives::types::{AccountId, EpochId, ShardId};
 
+use crate::chunk_validation::ChunkValidator;
 use crate::metrics;
 
 const CHUNK_HEADERS_FOR_INCLUSION_CACHE_SIZE: usize = 2048;
 const NUM_EPOCH_CHUNK_PRODUCERS_TO_KEEP_IN_BLOCKLIST: usize = 1000;
 
+struct ChunkInfo {
+    // chunk_header, received_time and chunk_producer are populated when we call mark_chunk_header_ready_for_inclusion
+    pub chunk_header: ShardChunkHeader,
+    pub received_time: DateTime<Utc>,
+    pub chunk_producer: AccountId,
+    // The signatures are populated later during call to chunk_validator.get_chunk_endorsement_signature
+    // This is a sort of cache to store previously fetched signatures.
+    pub signatures: Option<ChunkEndorsementSignatures>,
+}
```

Comment:
nit: it comments how the fields *work*, not what they *are*. I think it is better to put above `ChunkInfo` - say how the fields are populated - and then comment on each field what it represents.

And also, it doesn't seem that `signatures` are really a cache, based on implementation - it is just a set of signatures enough to be included. If new signatures arrive, it is not updated.

Reply:
It might make sense to remove this altogether in favor of fetching the signatures directly from `ChunkValidator` module... I don't think the "optimization" is worth the complexity. And like you mentioned, it's better to have a "prepare" step before calling `get_chunk_headers_ready_for_inclusion` that explicitly checks if we have enough endorsements.

Reply:
I personally just wasn't too happy with calling `get_chunk_endorsement_signatures` once for preparation and then once again for actually fetching it during block production.